treeherder/services/taskcluster.py

```diff
@@ -0,0 +1,137 @@
+        )
+
+    def _load(self, decision_task_id: str, task_id: str) -> dict:
+        if not decision_task_id:
+            raise ValueError("No decision task, can't find taskcluster actions")
+
+        # fetch
+        logger.debug('Fetching actions.json...')
+        actions_json = self.queue.getLatestArtifact(decision_task_id, 'public/actions.json')
+        task_definition = self.queue.task(task_id)
+
+        if actions_json['version'] != 1:
+            raise RuntimeError('Wrong version of actions.json, unable to continue')
+
+        return {
+            'staticActionVariables': actions_json['variables'],
+            'actions': self._filter_relevant_actions(actions_json, task_definition),
+        }
+
+    def _submit(self,
+                action=None,
+                decision_task_id=None,
+                task_id=None,
+                input=None,
+                static_action_variables=None) -> str:
+        context = {
+                "taskGroupId": decision_task_id,
+                "taskId": task_id or None,
+                "input": input,
+         }
+        context.update(static_action_variables)
+
+        if action["kind"] == "task":
+            raise NotImplementedError("Unable to submit actions with 'task' kind.")
+
+        if action["kind"] == "hook":
+            hook_payload = jsone.render(action["hookPayload"], context)
+            hook_id, hook_group_id = action["hookId"], action["hookGroupId"]
+
+            decision_task = self.queue.task(decision_task_id)
+            expansion = self.auth.expandScopes({"scopes": decision_task["scopes"]})
+            expression = f"in-tree:hook-action:{hook_group_id}/{hook_id}"
+
+            if not satisfiesExpression(expansion["scopes"], expression):
+                raise RuntimeError(f"Action is misconfigured: decision task's scopes do not satisfy {expression}")
+
+            result = self.hooks.triggerHook(hook_group_id, hook_id, hook_payload)
+            return result["status"]["taskId"]
+
+    @classmethod
+    def _filter_relevant_actions(cls, actions_json: dict, original_task) -> List:
+        relevant_actions = {}
+
+        for action in actions_json['actions']:
+            action_name = action['name']
+            if action_name in relevant_actions:
+                continue
+
+            no_context_or_task_to_check = (not len(action['context'])) and (not original_task)
+            task_is_in_context = (original_task and original_task.get('tags') and
+                                  cls._task_in_context(action['context'], original_task['tags']))
+
+            if no_context_or_task_to_check or task_is_in_context:
+                relevant_actions[action_name] = action
+
+        return list(relevant_actions.values())
+
+    @staticmethod
+    def _get_action(action_array: list, action_name: str) -> str:
+        try:
+            return [a for a in action_array if a["name"] == action_name][0]
```

Comment:
why are you returning [0], can you add a comment to indicate what you expect in index [0]?

Reply:
The expectation is that there will only be 0 or 1 item left in the list comprehension after this filter.  So this is equivalent to a `get` operation in Python, since there isn't one on `list`.  (Seems like there should be...  :) )

Another way to do this would be:

    action = next((a for a in action_array if a["name"] == action_name), None)

And then check for `action` having a value.  So this is just code style.  But maybe a comment would be helpful, too.